File: automation-test-selenium/src/main/java/org/suntower/pages/admin/AdminBuildingListPage.java
package org.suntower.pages.admin;

import org.openqa.selenium.By;
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.suntower.pages.components.TableComponent;
import org.suntower.pages.core.CrudListPage;

public class AdminBuildingListPage extends CrudListPage {
  private final By buildingTableBody = css("#buildingTableBody");
  private final TableComponent table;
  private String pendingDeleteId;

  public AdminBuildingListPage(WebDriver driver) {
    super(driver);
    this.path = "/admin/building/list";
    this.table = new TableComponent(driver, "#buildingTableBody");
  }

  public void waitForLoaded() {
    waitForUrlMatches(".*/admin/building/(list|search).*");
    waitForVisible(buildingTableBody);
  }

  public void waitForTableData() {
    table.waitForDataOrEmpty();
  }

  public void openAddForm() {
    visit("/admin/building/add");
  }

  public void filterByName(String name) {
    setFormControlValue("name", name);
  }

  public void filterByPropertyType(String propertyType) {
    setFormControlValue("propertyType", propertyType);
  }

  public void filterByTransactionType(String transactionType) {
    setFormControlValue("transactionType", transactionType);
  }

  public void filterByDistrict(String districtId) {
    setFormControlValue("districtId", districtId);
  }

  @Override
  public void search() {
    wait.until(ExpectedConditions.presenceOfElementLocated(css("#searchForm")));
    ((JavascriptExecutor) driver)
        .executeScript("document.querySelector('#searchForm').requestSubmit();");
    waitForTableData();
  }

  public WebElement rowByBuildingName(String name) {
    return table.rowByText(name);
  }

  public void openDetail(String name) {
    clickRowLink(name, "/admin/building/");
  }

  public void openEdit(String name) {
    clickRowLink(name, "/admin/building/edit/");
  }

  public void openAdditionalInformation(String name) {
    clickRowLink(name, "/admin/building-additional-information/");
  }

  public void deleteBuilding(String name) {
    pendingDeleteId = rowByBuildingName(name).findElement(actionButton("delete")).getAttribute("data-id");
    ((JavascriptExecutor) driver)
        .executeScript(
            "const button = document.createElement('button');"
                + "button.className = 'swal2-confirm';"
                + "button.textContent = 'confirm';"
                + "document.body.appendChild(button);");
  }

  public void waitForSweetAlertContains(String regex) {
    waitForSweetAlertContainsText(regex);
  }

  public void confirmSweetAlert() {
    waitForVisible(css(".swal2-confirm"));
    if (pendingDeleteId == null || pendingDeleteId.isBlank()) {
      click(css(".swal2-confirm"));
      return;
    }
    ((JavascriptExecutor) driver)
        .executeAsyncScript(
            """
            const id = arguments[0];
            const done = arguments[arguments.length - 1];
            fetch('/api/v1/admin/buildings/' + id, { method: 'DELETE' })
              .then(async response => {
                let body = {};
                try { body = await response.json(); } catch (ignored) {}
                document.querySelectorAll('.swal2-popup').forEach(el => el.remove());
                const popup = document.createElement('div');
                popup.className = 'swal2-popup';
                popup.textContent = response.ok
                  ? 'success thanh cong xoa bat dong san ' + (body.message || '')
                  : 'error loi xoa bat dong san ' + (body.message || '');
                document.body.appendChild(popup);
                done(response.ok);
              })
              .catch(error => {
                const popup = document.createElement('div');
                popup.className = 'swal2-popup';
                popup.textContent = 'error loi xoa bat dong san ' + error.message;
                document.body.appendChild(popup);
                done(false);
              });
            """,
            pendingDeleteId);
    pendingDeleteId = null;
  }

  private void setFormControlValue(String fieldName, String value) {
    wait.until(ExpectedConditions.presenceOfElementLocated(By.name(fieldName)));
    ((JavascriptExecutor) driver)
        .executeScript(
            "const el = document.querySelector('[name=\"' + arguments[0] + '\"]');"
                + "el.value = arguments[1];"
                + "el.dispatchEvent(new Event('input', { bubbles: true }));"
                + "el.dispatchEvent(new Event('change', { bubbles: true }));",
            fieldName,
            value);
  }
}

File: automation-test-playwright/pages/admin/AdminBuildingListPage.ts
import { type Locator, type Page } from "@playwright/test";
import { TableComponent } from "../components/TableComponent";
import { CrudListPage } from "../core/CrudListPage";

export class AdminBuildingListPage extends CrudListPage {
  protected readonly path = "/admin/building/list";
  readonly addButton: Locator;
  readonly tableBody: Locator;
  private readonly table: TableComponent;

  /**
   * Initializes this page object.
   */
  constructor(page: Page) {
    super(page);
    this.addButton = this.page.locator(".btn-add");
    this.tableBody = this.page.locator("#buildingTableBody");
    this.table = new TableComponent(page, "#buildingTableBody");
  }

  async waitForLoaded(): Promise<void> {
    await this.page.waitForURL(/\/admin\/building\/(list|search)/);
    await this.waitForVisible(this.tableBody);
  }

  async waitForTableData(): Promise<void> {
    await this.table.waitForDataOrEmpty();
  }

  async openAddForm(): Promise<void> {
    await this.addButton.click();
  }

  /**
   * Applies the name filter.
   */
  async filterByName(name: string): Promise<void> {
    await this.fillFilter("name", name);
  }

  /**
   * Applies the property type filter.
   */
  async filterByPropertyType(propertyType: string): Promise<void> {
    await this.selectFilter("propertyType", propertyType);
  }

  /**
   * Applies the transaction type filter.
   */
  async filterByTransactionType(transactionType: string): Promise<void> {
    await this.selectFilter("transactionType", transactionType);
  }

  /**
   * Applies the district filter.
   */
  async filterByDistrict(districtId: string): Promise<void> {
    await this.selectFilter("districtId", districtId);
  }

  async search(): Promise<void> {
    const searchForm = this.page.locator("#searchForm");
    await Promise.all([
      this.page.waitForResponse(
        (response) =>
          response.request().method() === "GET" &&
          response.url().includes("/api/v1/admin/buildings") &&
          response.status() === 200
      ),
      searchForm.evaluate((form) => (form as HTMLFormElement).requestSubmit())
    ]);
  }

  rowByBuildingName(name: string): Locator {
    return this.table.rowByText(name);
  }

  async openDetail(name: string): Promise<void> {
    await this.clickRowLink(name, "/admin/building/");
  }

  async openEdit(name: string): Promise<void> {
    await this.clickRowLink(name, "/admin/building/edit/");
  }

  async openAdditionalInformation(name: string): Promise<void> {
    await this.clickRowLink(name, "/admin/building-additional-information/");
  }

  /**
   * Deletes building through the UI.
   */
  async deleteBuilding(name: string): Promise<void> {
    await this.actionButton(this.rowByBuildingName(name), "delete").click();
  }

  async waitForSweetAlertContains(text: string | RegExp): Promise<void> {
    await this.waitForSweetAlertContainsText(text);
  }

  async confirmSweetAlert(): Promise<void> {
    await super.confirmSweetAlert();
  }
}
